### ISSUE TITLE
Multiple y axes

### ISSUE THREAD

**Issue by drmcnelson** [JamesBremner/windex#25]
Hi,

I want to add a second static trace, and be able to make it appear or not appear in the graph.

Would calling clear() be a suitable way to make the trace not appear in the graph?

Thank you

**Comment by JamesBremner** [comment]
That might well work, but it is clumsy since you would have to add data again when you wanted it to reappear.

Should be easy enough to add some call to hide/display a trace.  

Maybe 

    wex::plot::trace::show( bool f = true );

**Comment by JamesBremner** [comment]
Much simpler:

call wex::plot::trace::color() and set the color to the plot background you are using.

No changes to windex required

**Comment by drmcnelson** [comment]
Will be trying this shortly,  I am thinking to use show().

**Comment by JamesBremner** [comment]
Do not use show. It will not work.

Please read my comment https://github.com/JamesBremner/windex/issues/25#issuecomment-1961793712

wex::plot::trace::color() is the way to go.

**Comment by drmcnelson** [comment]
What happens when the trace is cleared?   Initially they might be empty unless I give them dummy data.

**Comment by JamesBremner** [comment]
This has not moved in a week!  Can we resolve it?

**Comment by drmcnelson** [comment]
Oh, I worked around it by simply sticking with one trace for now.  I might come back to it later in the week.  

 It seems complicated for the moment, when the traces are all similar magnitude it is an easy concept, when they differ by orders of magnitude, people usually use multiple y axes.

I'll say, how about we close it for now.  I will re-open if I start to use it or want to contribute something.

**Comment by JamesBremner** [comment]
Did you try my suggestion?  https://github.com/JamesBremner/windex/issues/25#issuecomment-1961793712

**Comment by JamesBremner** [comment]
Did you try my suggestion? https://github.com/JamesBremner/windex/issues/25#issuecomment-1961793712

**Comment by drmcnelson** [comment]
Two questions

1)  Does auto scaling consider all of the traces?  

2)   Does the second trace increase the execution time for updates by a significant amount?

So, my concern is that if both answers are yes, then simply coloring it to background will leave a problem.  In some instances the traces will span very different ranges in magnitude. 

To give some specificity to this, consider an absorption spectrum.   See the picture below for the setup.     For reference, here is the equation.  The absorption is the log of the ratio of two spectra

      Absorbance = log10(  incident light / transmitted light ).  

In practice, absorbances are usually reported in the range from 0 to 2 (which corresponds to the transmitted light being 100% to 1% of the incident light).

The raw spectra go from 0 to 0.6Volts, or as many like to report it, 0 to 65,535 "counts" for the binary readout.  (I know, the latter is absurd bordering on meaningless, but that is the way many report it, even those who should know better).

Thus, the the traces differ by either a factor of 4, or, a factor of 32,000, depending on which way the raw spectra are presented.

So, let's assume auto scaling looks at all three.   As a result, the trace spanning the lower range will not be shown in a useful way, even when the others are colored to background.

![IMG_20240308_095618042 p400](https://github.com/JamesBremner/windex/assets/38619857/addf19b3-c558-4de2-aee1-7481006a44dd)

**Comment by drmcnelson** [comment]
P/S the way most deal with this, is by making the y axis to the right of the graph a different scale, as a second y axis.   

At the moment I am only showing one trace and even with that, I did not implement the control yet to load the trace from the dark or reference spectra.

**Comment by JamesBremner** [comment]
I need to point out that the wex::plot does not have a feature called 'autoscaling'  I assume you mean fit scaling, where the plot displayed covers the entire and only the entire range of the input data.

-------------------------------

To answer your questions

> Does fit scaling consider all of the traces?

Yes.

> Does the second trace increase the execution time for updates by a significant amount?

No.  Finding the data bounds for each trace takes just a few millisecs.

------------------------------------

Your questions have no relevance to your actual concern, which is valid.

> Thus, the the traces differ by either a factor of 4, or, a factor of 32,000, depending on which way the raw spectra are presented.
> So, let's assume auto scaling looks at all three. As a result, the trace spanning the lower range will not be shown in a useful way, even when the others are colored to background.

More succinctly:  Traces with very different data ranges cannot be shown on the same plot - the trace with the smaller data range will be reduced to an uninformative horizontal line.

----------------------------------------

Bottom line: You want an option to remove/include individual traces from the scaling calculations  ( not just hide them, but exclude them from the scale calculation )

------------------------------------------

What next:

- Edit issue title so that it describes what you actually want
- Label with 'enhancement'
- Assign issue to me.

-------------------------------------------------

Possible future work:

You might want to consider asking for traces to be individually scaled, so they can be shown on the same plot each with their own scale.

( This is a major enhancement, not just cosmetic.  We would need to discuss financial support for such an enhancement )

**Comment by drmcnelson** [comment]
A few milliseconds per trace, can indeed be very much a concern.   Maybe not so much for the present application.

If it is only scaling, I can do that in the data.  A second y-axis with its own tic marks and labels would be good.  And, a "legend" box would be useful (a table of line sample and label) so that we can tell the viewer which trace is which.

Yes, I guessed it is a lot of work, and, I do have a perfectly okay work around for the time being, and, even if available, I have to finish up some other work before I can come back to it.

I don't want to ask you to do this.  It is too much.

**Comment by JamesBremner** [comment]
So close the issue

**Comment by drmcnelson** [comment]
While we are on the subject,  axis labels would be useful.

By axis labels, I mean something like "Wavelength" below the x axis tic labels and "Volts" to the left of the y axis tic labels.

**Comment by drmcnelson** [comment]
Okay

**Comment by JamesBremner** [comment]
You can add labels wherever you want in the usual way

`wex::maker::make<wex::label>(plot);`

**Comment by drmcnelson** [comment]
Yes, but to put them below the tics of the x axis, or the left of the y axis, you need more space.

So that means a setting for the margins inside of the plot area.   Do we have that?

**Comment by JamesBremner** [comment]
The margins are hard coded.  Please open an enhancement issue for this and **_assign it to me_**

**Comment by drmcnelson** [comment]
Okay, one more related question,  is it possible to nest a form within a form?

**Comment by JamesBremner** [comment]
Depends on what exactly you mean by "nesting".  Open a question issue to discuss this in detail.
